// appium-ios-remotexpc/build/src/services/ios/dvt/instruments/graphics.js
import { getLogger } from '../../../../lib/logger.js';
import { MessageAux } from '../dtx-message.js';
const log = getLogger('Graphics');
export class Graphics {
    dvt;
    static IDENTIFIER = 'com.apple.instruments.server.services.graphics.opengl';
    channel = null;
    constructor(dvt) {
        this.dvt = dvt;
    }
    async initialize() {
        if (!this.channel) {
            this.channel = await this.dvt.makeChannel(Graphics.IDENTIFIER);
        }
    }
    async start() {
        await this.initialize();
        const args = new MessageAux().appendObj(0.0);
        await this.channel.call('startSamplingAtTimeInterval_')(args);
        await this.channel.receivePlist();
    }
    async stop() {
        await this.channel.call('stopSampling')();
    }
    async *messages() {
        log.debug('logging started');
        await this.start();
        try {
            while (true) {
                yield await this.channel.receivePlist();
            }
        }
        finally {
            log.debug('logging stopped');
            await this.stop();
        }
    }
}
