// appium-ios-remotexpc/src/index.ts
import { STRONGBOX_CONTAINER_NAME } from './constants.js';
import { createLockdownServiceByUDID } from './lib/lockdown/index.js';
import {
  PacketStreamClient,
  PacketStreamServer,
  TunnelManager,
} from './lib/tunnel/index.js';
import {
  TunnelRegistryServer,
  startTunnelRegistryServer,
} from './lib/tunnel/tunnel-registry-server.js';
import { Usbmux, createUsbmux } from './lib/usbmux/index.js';
import * as Services from './services.js';
import { startCoreDeviceProxy } from './services/ios/tunnel-service/index.js';

export type {
  DiagnosticsService,
  MobileImageMounterService,
  NotificationProxyService,
  MobileConfigService,
  PowerAssertionService,
  PowerAssertionOptions,
  SpringboardService,
  WebInspectorService,
  MisagentService,
  SyslogService,
  DVTSecureSocketProxyService,
  LocationSimulationService,
  ConditionInducerService,
  ScreenshotService,
  GraphicsService,
  DeviceInfoService,
  ProcessInfo,
  ConditionProfile,
  ConditionGroup,
  SocketInfo,
  TunnelResult,
  TunnelRegistry,
  TunnelRegistryEntry,
  DiagnosticsServiceWithConnection,
  MobileImageMounterServiceWithConnection,
  NotificationProxyServiceWithConnection,
  MobileConfigServiceWithConnection,
  PowerAssertionServiceWithConnection,
  SpringboardServiceWithConnection,
  WebInspectorServiceWithConnection,
  MisagentServiceWithConnection,
  DVTServiceWithConnection,
} from './lib/types.js';
export { PowerAssertionType } from './lib/types.js';
export {
  STRONGBOX_CONTAINER_NAME,
  createUsbmux,
  Services,
  Usbmux,
  TunnelManager,
  PacketStreamServer,
  PacketStreamClient,
  createLockdownServiceByUDID,
  startCoreDeviceProxy,
  TunnelRegistryServer,
  startTunnelRegistryServer,
};
