// appium-ios-remotexpc/build/src/services/ios/afc/enums.js
/**
 * Enums for AFC protocol as defined by Apple's AFC service
 */
export var AfcOpcode;
(function (AfcOpcode) {
    AfcOpcode[AfcOpcode["STATUS"] = 1] = "STATUS";
    AfcOpcode[AfcOpcode["DATA"] = 2] = "DATA";
    AfcOpcode[AfcOpcode["READ_DIR"] = 3] = "READ_DIR";
    AfcOpcode[AfcOpcode["REMOVE_PATH"] = 8] = "REMOVE_PATH";
    AfcOpcode[AfcOpcode["MAKE_DIR"] = 9] = "MAKE_DIR";
    AfcOpcode[AfcOpcode["GET_FILE_INFO"] = 10] = "GET_FILE_INFO";
    AfcOpcode[AfcOpcode["GET_DEVINFO"] = 11] = "GET_DEVINFO";
    AfcOpcode[AfcOpcode["FILE_OPEN"] = 13] = "FILE_OPEN";
    AfcOpcode[AfcOpcode["READ"] = 15] = "READ";
    AfcOpcode[AfcOpcode["WRITE"] = 16] = "WRITE";
    AfcOpcode[AfcOpcode["FILE_CLOSE"] = 20] = "FILE_CLOSE";
    AfcOpcode[AfcOpcode["RENAME_PATH"] = 24] = "RENAME_PATH";
    AfcOpcode[AfcOpcode["MAKE_LINK"] = 28] = "MAKE_LINK";
})(AfcOpcode || (AfcOpcode = {}));
export var AfcError;
(function (AfcError) {
    AfcError[AfcError["SUCCESS"] = 0] = "SUCCESS";
    AfcError[AfcError["UNKNOWN_ERROR"] = 1] = "UNKNOWN_ERROR";
    AfcError[AfcError["OP_HEADER_INVALID"] = 2] = "OP_HEADER_INVALID";
    AfcError[AfcError["NO_RESOURCES"] = 3] = "NO_RESOURCES";
    AfcError[AfcError["READ_ERROR"] = 4] = "READ_ERROR";
    AfcError[AfcError["WRITE_ERROR"] = 5] = "WRITE_ERROR";
    AfcError[AfcError["UNKNOWN_PACKET_TYPE"] = 6] = "UNKNOWN_PACKET_TYPE";
    AfcError[AfcError["INVALID_ARG"] = 7] = "INVALID_ARG";
    AfcError[AfcError["OBJECT_NOT_FOUND"] = 8] = "OBJECT_NOT_FOUND";
    AfcError[AfcError["OBJECT_IS_DIR"] = 9] = "OBJECT_IS_DIR";
    AfcError[AfcError["PERM_DENIED"] = 10] = "PERM_DENIED";
    AfcError[AfcError["SERVICE_NOT_CONNECTED"] = 11] = "SERVICE_NOT_CONNECTED";
    AfcError[AfcError["OP_TIMEOUT"] = 12] = "OP_TIMEOUT";
    AfcError[AfcError["TOO_MUCH_DATA"] = 13] = "TOO_MUCH_DATA";
    AfcError[AfcError["END_OF_DATA"] = 14] = "END_OF_DATA";
    AfcError[AfcError["OP_NOT_SUPPORTED"] = 15] = "OP_NOT_SUPPORTED";
    AfcError[AfcError["OBJECT_EXISTS"] = 16] = "OBJECT_EXISTS";
    AfcError[AfcError["OBJECT_BUSY"] = 17] = "OBJECT_BUSY";
    AfcError[AfcError["NO_SPACE_LEFT"] = 18] = "NO_SPACE_LEFT";
    AfcError[AfcError["OP_WOULD_BLOCK"] = 19] = "OP_WOULD_BLOCK";
    AfcError[AfcError["IO_ERROR"] = 20] = "IO_ERROR";
    AfcError[AfcError["OP_INTERRUPTED"] = 21] = "OP_INTERRUPTED";
    AfcError[AfcError["OP_IN_PROGRESS"] = 22] = "OP_IN_PROGRESS";
    AfcError[AfcError["INTERNAL_ERROR"] = 23] = "INTERNAL_ERROR";
    AfcError[AfcError["MUX_ERROR"] = 30] = "MUX_ERROR";
    AfcError[AfcError["NO_MEM"] = 31] = "NO_MEM";
    AfcError[AfcError["NOT_ENOUGH_DATA"] = 32] = "NOT_ENOUGH_DATA";
    AfcError[AfcError["DIR_NOT_EMPTY"] = 33] = "DIR_NOT_EMPTY";
})(AfcError || (AfcError = {}));
export var AfcLinkType;
(function (AfcLinkType) {
    AfcLinkType[AfcLinkType["HARDLINK"] = 1] = "HARDLINK";
    AfcLinkType[AfcLinkType["SYMLINK"] = 2] = "SYMLINK";
})(AfcLinkType || (AfcLinkType = {}));
export var AfcFopenMode;
(function (AfcFopenMode) {
    AfcFopenMode[AfcFopenMode["RDONLY"] = 1] = "RDONLY";
    AfcFopenMode[AfcFopenMode["RW"] = 2] = "RW";
    AfcFopenMode[AfcFopenMode["WRONLY"] = 3] = "WRONLY";
    AfcFopenMode[AfcFopenMode["WR"] = 4] = "WR";
    AfcFopenMode[AfcFopenMode["APPEND"] = 5] = "APPEND";
    AfcFopenMode[AfcFopenMode["RDAPPEND"] = 6] = "RDAPPEND";
})(AfcFopenMode || (AfcFopenMode = {}));
export var AfcFileMode;
(function (AfcFileMode) {
    AfcFileMode["S_IFDIR"] = "S_IFDIR";
    AfcFileMode["S_IFREG"] = "S_IFREG";
    AfcFileMode["S_IFLNK"] = "S_IFLNK";
})(AfcFileMode || (AfcFileMode = {}));
