// appium-ios-remotexpc/build/src/lib/apple-tv/storage/pairing-storage.js
import { strongbox } from '@appium/strongbox';
import { logger } from '@appium/support';
import { STRONGBOX_CONTAINER_NAME } from '../../../constants.js';
import { createXmlPlist } from '../../plist/index.js';
import { PairingError } from '../errors.js';
/** Manages persistent storage of pairing credentials as plist files */
export class PairingStorage {
    config;
    log = logger.getLogger('PairingStorage');
    box;
    constructor(config) {
        this.config = config;
        this.box = strongbox(STRONGBOX_CONTAINER_NAME);
    }
    async save(deviceId, ltpk, ltsk, remoteUnlockHostKey = '') {
        try {
            const itemName = `appletv_pairing_${deviceId}`;
            const plistContent = this.createPlistContent(ltpk, ltsk, remoteUnlockHostKey);
            const item = await this.box.createItemWithValue(itemName, plistContent);
            const itemPath = item.id;
            this.log.info(`Pairing record saved to: ${itemPath}`);
            return itemPath;
        }
        catch (error) {
            this.log.error('Save pairing record error:', error);
            throw new PairingError('Failed to save pairing record', 'SAVE_ERROR', error);
        }
    }
    createPlistContent(publicKey, privateKey, remoteUnlockHostKey) {
        return createXmlPlist({
            private_key: privateKey,
            public_key: publicKey,
            remote_unlock_host_key: remoteUnlockHostKey,
        });
    }
}
